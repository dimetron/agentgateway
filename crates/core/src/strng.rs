use std::fmt::Error;
use std::ops::Deref;

use arcstr::ArcStr;
use prometheus_client::encoding::{LabelKeyEncoder, LabelValueEncoder};

/// 'Strng' provides a string type that has better properties for our use case:
/// * Cheap cloning (ref counting)
/// * Efficient storage (8 bytes vs 24 bytes)
/// * Immutable
///
/// This is mostly provided by a library, ArcStr, we just provide a very thin wrapper around it
/// for some flexibility.
pub type Strng = ArcStr;

pub const EMPTY: Strng = literal!("");

pub fn new<A: AsRef<str>>(s: A) -> Strng {
	Strng::from(s.as_ref())
}

pub use arcstr::{format, literal};

/// RichStrng wraps Strng to let us implement arbitrary methods. How annoying.
#[derive(Clone, Hash, Default, Debug, PartialEq, Eq)]
pub struct RichStrng(Strng);

impl prometheus_client::encoding::EncodeLabelValue for RichStrng {
	fn encode(&self, encoder: &mut LabelValueEncoder) -> Result<(), Error> {
		prometheus_client::encoding::EncodeLabelValue::encode(
			&<ArcStr as AsRef<str>>::as_ref(&self.0),
			encoder,
		)
	}
}

impl prometheus_client::encoding::EncodeLabelKey for RichStrng {
	fn encode(&self, encoder: &mut LabelKeyEncoder) -> Result<(), Error> {
		prometheus_client::encoding::EncodeLabelKey::encode(
			&<ArcStr as AsRef<str>>::as_ref(&self.0),
			encoder,
		)
	}
}

impl Deref for RichStrng {
	type Target = Strng;

	fn deref(&self) -> &Self::Target {
		&self.0
	}
}

impl<T> From<T> for RichStrng
where
	T: Into<Strng>,
{
	fn from(value: T) -> Self {
		RichStrng(value.into())
	}
}

#[cfg(test)]
mod test {
	use super::*;
	fn as_ref_fn<A: AsRef<str>>(_s: A) {}
	fn into_string_fn<A: Into<String>>(_s: A) {}
	fn string_fn(_s: String) {}
	fn str_fn(_s: &str) {}

	#[test]
	fn interning() {
		// Mostly we just thinly wrap ArcString, so just validate our assumptions about the library
		let a = new("abc");
		let b = new("abc");
		assert_eq!(std::mem::size_of::<Strng>(), 8);
		assert_eq!(std::format!("{a}"), "abc");
		assert_eq!(super::format!("{a}"), "abc");
		assert_eq!(ArcStr::strong_count(&a), ArcStr::strong_count(&b));
		assert_eq!(ArcStr::strong_count(&a), Some(1));
		let c = a.clone();
		assert_eq!(ArcStr::strong_count(&a), ArcStr::strong_count(&c));
		assert_eq!(ArcStr::strong_count(&a), Some(2));
		assert_eq!("abc", b.to_string());

		// Compile time assertion we can call function in various ways
		as_ref_fn(new("abc"));
		into_string_fn(&*new("abc"));
		string_fn(a.to_string());
		str_fn(&new("abc"));
	}
}
