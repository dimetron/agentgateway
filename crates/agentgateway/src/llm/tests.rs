use std::fs;
use std::path::Path;

use agent_core::strng;
use serde::de::DeserializeOwned;
use serde_json::Value;

use super::*;

fn test_response<T: DeserializeOwned>(
	test_name: &str,
	xlate: impl Fn(T) -> Result<universal::Response, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let provider_str = &fs::read_to_string(&input_path)
		.unwrap_or_else(|_| panic!("{test_name}: Failed to read input file"));
	let provider_raw: Value = serde_json_path_to_error::from_str(provider_str)
		.unwrap_or_else(|_| panic!("{test_name}: Failed to parse provider json"));
	let provider: T = serde_json_path_to_error::from_str(provider_str)
		.unwrap_or_else(|_| panic!("{test_name}: Failed to parse provider JSON"));

	let openai_response =
		xlate(provider).expect("Failed to translate provider response to OpenAI format");

	insta::with_settings!({
			info => &provider_raw,
			description => input_path.to_string_lossy().to_string(),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(test_name, openai_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

fn test_request<T: Serialize>(
	provider_name: &str,
	test_name: &str,
	xlate: impl Fn(universal::Request) -> Result<T, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let openai_str = &fs::read_to_string(&input_path).expect("Failed to read input file");
	let openai_raw: Value = serde_json::from_str(openai_str).expect("Failed to parse openai json");
	let openai: universal::Request =
		serde_json::from_str(openai_str).expect("Failed to parse openai JSON");

	let provider_response =
		xlate(openai).expect("Failed to translate OpenAI format to provider request ");

	insta::with_settings!({
			info => &openai_raw,
			description => format!("{}: {}", provider_name, test_name),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(format!("{}-{}", provider_name, test_name), provider_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

const ALL_REQUESTS: &[&str] = &["request_basic", "request_full", "request_tool-call"];

#[test]
fn test_bedrock() {
	let response = |i| bedrock::translate_response(i, &strng::new("fake-model"));
	test_response::<bedrock::types::ConverseResponse>("response_bedrock_basic", response);
	test_response::<bedrock::types::ConverseResponse>("response_bedrock_tool", response);
	let provider = bedrock::Provider {
		model: Some(strng::new("test-model")),
		region: strng::new("us-east-1"),
		guardrail_identifier: None,
		guardrail_version: None,
	};
	let request = |i| Ok(bedrock::translate_request(i, &provider));
	for r in ALL_REQUESTS {
		test_request("bedrock", r, request);
	}
}

#[test]
fn test_anthropic() {
	let response = |i| Ok(anthropic::translate_response(i));
	test_response::<anthropic::types::MessagesResponse>("response_anthropic_basic", response);
	test_response::<anthropic::types::MessagesResponse>("response_anthropic_tool", response);

	let request = |i| Ok(anthropic::translate_request(i));
	for r in ALL_REQUESTS {
		test_request("anthropic", r, request);
	}
}
