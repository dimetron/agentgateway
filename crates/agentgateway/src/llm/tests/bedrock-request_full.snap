---
source: crates/agentgateway/src/llm/tests.rs
description: "bedrock: request_full"
info:
  model: gpt-4-turbo-preview
  messages:
    - role: system
      content: You are a helpful assistant that specializes in technical documentation and code analysis.
      name: system_assistant
    - role: user
      content: "Please analyze this code snippet and explain what it does: function fibonacci(n) { return n <= 1 ? n : fibonacci(n-1) + fibonacci(n-2); }"
      name: user_1
    - role: assistant
      content: This is a recursive implementation of the Fibonacci sequence. It calculates the nth Fibonacci number by adding the previous two numbers.
      name: assistant_1
    - role: user
      content: Can you optimize it for better performance?
      name: user_2
  max_tokens: 1000
  temperature: 0.7
  top_p: 0.9
  n: 3
  stream: false
  stop:
    - "\n\n"
    - END
  presence_penalty: 0.1
  frequency_penalty: 0.2
  logit_bias:
    "50256": -100
  user: user_12345
  response_format:
    type: json_object
  seed: 42
  tool_choice:
    type: function
    function:
      name: my_function
  tools:
    - type: function
      function:
        name: get_weather
        description: Get the current weather in a given location
        parameters:
          type: object
          properties:
            location:
              type: string
              description: "The city and state, e.g. San Francisco, CA"
            unit:
              type: string
              enum:
                - celsius
                - fahrenheit
          required:
            - location
  functions:
    - name: calculate_fibonacci
      description: Calculate the nth Fibonacci number using an optimized algorithm
      parameters:
        type: object
        properties:
          n:
            type: integer
            description: The position in the Fibonacci sequence (0-indexed)
            minimum: 0
            maximum: 100
        required:
          - n
  function_call: auto
---
{
  "modelId": "gpt-4-turbo-preview",
  "messages": [
    {
      "role": "user",
      "content": [
        {
          "text": "Please analyze this code snippet and explain what it does: function fibonacci(n) { return n <= 1 ? n : fibonacci(n-1) + fibonacci(n-2); }"
        }
      ]
    },
    {
      "role": "assistant",
      "content": [
        {
          "text": "This is a recursive implementation of the Fibonacci sequence. It calculates the nth Fibonacci number by adding the previous two numbers."
        }
      ]
    },
    {
      "role": "user",
      "content": [
        {
          "text": "Can you optimize it for better performance?"
        }
      ]
    }
  ],
  "system": [
    {
      "text": "You are a helpful assistant that specializes in technical documentation and code analysis."
    }
  ],
  "inferenceConfig": {
    "maxTokens": 1000,
    "temperature": 0.7,
    "top_p": 0.9,
    "stopSequences": [
      "\n\n",
      "END"
    ]
  },
  "toolConfig": {
    "tools": [
      {
        "toolSpec": {
          "name": "get_weather",
          "description": "Get the current weather in a given location",
          "inputSchema": {
            "json": {
              "type": "object",
              "properties": {
                "location": {
                  "type": "string",
                  "description": "The city and state, e.g. San Francisco, CA"
                },
                "unit": {
                  "type": "string",
                  "enum": [
                    "celsius",
                    "fahrenheit"
                  ]
                }
              },
              "required": [
                "location"
              ]
            }
          }
        }
      }
    ],
    "tool_choice": {
      "tool": {
        "name": "my_function"
      }
    }
  },
  "requestMetadata": {
    "user_id": "user_12345"
  }
}
